neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/impl/BoltNeo4jConnectionImpl.java

```diff
@@ -277,31 +276,30 @@ private void setBookmark() throws SQLClientInfoException {
 	@Override public void close() throws SQLException {
 		try {
 			if (!this.isClosed()) {
-				if (this.transaction != null) {
-					this.transaction.close();
-					this.transaction = null;
-				}
-                if (this.session != null) {
-                    this.session.close();
-                    this.session = null;
-                    this.driver.close();
-                    this.driver = null;
-                }
-            }
+				closeTransaction();
+				closeSession(true);
+			}
 		} catch (Exception e) {
 			throw new SQLException("A database access error has occurred: " + e.getMessage());
 		}
 	}
 
-	/**
-	 * If has used {@link #newNeo4jSession()}
-	 * @param session
-	 */
-	public void closeSession(Session session){
-		if (session != null) {
-			session.close();
+	private void closeSession(boolean closeDriver) {
+		if (this.session != null) {
```

Comment:
what about error handling when session closing fails? then we should still try to close the driver if the parameter is true
and we should probably still set it to null so it gets GCed and the other logic an create new sessions.
same for closing the driver.

Reply:
Added safe close management.
We set it to null at line 291